keps/draft-20190503-framework-testing.md

```diff
@@ -0,0 +1,347 @@
+
+kep-number: draft-20190503
+title: Framework Testing
+authors:
+  - "@jbarrick-mesosphere"
+owners:
+  - "TBD"
+editor: TBD
+creation-date: 2019-05-03
+last-updated: 2019-05-03
+status: provisional
+see-also:
+  - KEP-0004
+---
+
+# Framework Testing
+
+## Table of Contents
+
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+     * [Goals](#goals)
+     * [Non-Goals](#non-goals)
+  * [User Stories](#user-stories)
+     * [KUDO Developers](#kudo-developers)
+     * [Framework Developer](#framework-developer)
+     * [Cluster Administrator](#cluster-administrator)
+     * [Application Operator](#application-operator)
+  * [Risks and Mitigations](#risks-and-mitigations)
+  * [Proposal](#proposal)
+     * [Definitions](#definitions)
+     * [Running the tests](#running-the-tests)
+     * [Directory structure](#directory-structure)
+     * [Test case file structure](#test-case-file-structure)
+        * [TestCase object](#testcase-object)
+        * [TestAssert object](#testassert-object)
+        * [Test constraints](#test-constraints)
+     * [Results collection](#results-collection)
+     * [Garbage collection](#garbage-collection)
+  * [Graduation Criteria](#graduation-criteria)
+  * [Implementation History](#implementation-history)
+  * [Alternatives](#alternatives)
+  * [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+In order to ensure the reliability of frameworks built on KUDO, it is important that developers of Frameworks are able
+to simply author and run tests that validate that their frameworks are running correctly. Tests can be run in CI to
+verify changes, by Framework developers in their development cycle, and by cluster admnistrators to verify that
+Frameworks are fully functional in their Kubernetes clusters, across a variety of configurations.
+
+This document outlines a simple, declarative framework for authoring and running acceptance tests for KUDO
+frameworks.
+
+## Motivation
+
+### Goals
+
+* Allow Framework developers to write acceptance and unit tests for Frameworks.
+* Run tests for Frameworks on pull requests for Frameworks.
```

Comment:
This is a good question. Similar, who is responsible for ensuring that the tests written are not flaky? Possibly framework owners, but it could be difficult to motivate their attention.

But as far as ownership, we could potentially leverage Github [CODEOWNERS](https://help.github.com/en/articles/about-code-owners) to enforce review by any set of people for both tests and framework changes?

Reply:
I'm not totally sure I follow the questions, is it unclear taken in context? The goal immediately above it says: `Allow Framework developers to write acceptance and unit tests for Frameworks.`.